webhooks/cluster.go

```diff
@@ -52,54 +58,191 @@ func (c *Cluster) Default(_ context.Context, obj runtime.Object) error {
 	}
-
-	// Note: The code in ValidateCreate is intentionally not duplicated to avoid that we accidentally
-	// implement new checks in the API package and forget to duplicate them to the webhook package.
-	// The idea is to add new validation which requires a reader and Cluster variable/patch validation
-	// in the webhook package.
-	// When we drop the method in the API package we must inline it here.
-	return cluster.ValidateCreate()
+	return c.validate(nil, cluster)
 }
 
-// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+// ValidateUpdate implements webhook.CustomValidator so a webhook will be registered for the type.
 func (c *Cluster) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) error {
-	cluster, ok := newObj.(*clusterv1.Cluster)
+	newCluster, ok := newObj.(*clusterv1.Cluster)
 	if !ok {
 		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", newObj))
 	}
 	oldCluster, ok := oldObj.(*clusterv1.Cluster)
 	if !ok {
 		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", oldObj))
 	}
+	return c.validate(oldCluster, newCluster)
+}
 
-	// Note: The code in ValidateUpdate is intentionally not duplicated to avoid that we accidentally
-	// implement new checks in the API package and forget to duplicate them to the webhook package.
-	// The idea is to add new validation which requires a reader and Cluster variable/patch validation
-	// in the webhook package.
-	// When we drop the method in the API package we must inline it here.
-	return cluster.ValidateUpdate(oldCluster)
+// ValidateDelete implements webhook.CustomValidator so a webhook will be registered for the type.
+func (c *Cluster) ValidateDelete(_ context.Context, obj runtime.Object) error {
+	return nil
 }
 
-// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
-func (c *Cluster) ValidateDelete(ctx context.Context, obj runtime.Object) error {
-	cluster, ok := obj.(*clusterv1.Cluster)
-	if !ok {
-		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", obj))
+func (c *Cluster) validate(old, new *clusterv1.Cluster) error {
+	var allErrs field.ErrorList
+	if new.Spec.InfrastructureRef != nil && new.Spec.InfrastructureRef.Namespace != new.Namespace {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				field.NewPath("spec", "infrastructureRef", "namespace"),
+				new.Spec.InfrastructureRef.Namespace,
+				"must match metadata.namespace",
+			),
+		)
+	}
+
+	if new.Spec.ControlPlaneRef != nil && new.Spec.ControlPlaneRef.Namespace != new.Namespace {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				field.NewPath("spec", "controlPlaneRef", "namespace"),
+				new.Spec.ControlPlaneRef.Namespace,
+				"must match metadata.namespace",
+			),
+		)
+	}
+
+	// Validate the managed topology, if defined.
+	if new.Spec.Topology != nil {
+		if topologyErrs := c.validateTopology(old, new); len(topologyErrs) > 0 {
```

Comment:
```suggestion
		if topologyErrs := validateTopology(old, new); len(topologyErrs) > 0 {
```

I'm not sure, but I think I would prefer making this a simple package func instead of a method. (similar for a few other  methods (also on ClusterClass)). This avoids having a `c` variable which is of type webhook.Cluster which is slightly confusing.

Note: we should only do this for methods which don't require access to c.Client later (but I think that at most a few top-level methods like c.validate need the Client)

Reply:
This is a matter of namespacing more than anything - because we're going to add a number of webhooks in the same package. validateTopology is a name that we might want to use again so I've added a receiver instead of  making it validateClusterTopology - I've no issue changing it if that's preferred though.

Reply:
That's a very very good point. In that case I would probably try to make everything which belongs to a webhook and it's not very clear that we want to reuse it elsewhere a method. (so the global namespace becomes an explicit opt-in)

But feel free to leave it as is.

Reply:
I think that makes sense. Splitting the difference like I've done here is confusing as it's not clear what func has a receiver for what reason. Shared functions, if we have any, can probably sit in their own file.

Reply:
Sounds good!